File: src/cleansing/pii_detector.py
"""
PII Detection and Masking Module
Detects and masks personally identifiable information from text
"""
import re
from typing import List, Dict, Tuple


class PIIDetector:
    """Detects and masks various types of PII in text"""
    
    def __init__(self):
        # Regex patterns for different PII types
        self.patterns = {
            'email': r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b',
            'phone': r'\b(?:\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b',
            'ssn': r'\b\d{3}-\d{2}-\d{4}\b',
            'credit_card': r'\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b',
            'ip_address': r'\b(?:\d{1,3}\.){3}\d{1,3}\b',
            'date': r'\b\d{1,2}[-/]\d{1,2}[-/]\d{2,4}\b',
            'address': r'\b\d+\s+[A-Za-z0-9\s,]+(?:Street|St|Avenue|Ave|Road|Rd|Boulevard|Blvd|Lane|Ln|Drive|Dr)\b',
        }
        
        # Common names patterns (simplified)
        self.name_patterns = [
            r'\b[A-Z][a-z]+\s+[A-Z][a-z]+\b',  # First Last
            r'\b(?:Mr\.|Mrs\.|Ms\.|Dr\.)\s+[A-Z][a-z]+\b',  # Title Name
        ]
    
    def detect_pii(self, text: str) -> List[Dict[str, any]]:
        """
        Detect PII in text and return findings
        
        Args:
            text: Input text to analyze
            
        Returns:
            List of dictionaries containing PII type, value, and position
        """
        findings = []
        
        for pii_type, pattern in self.patterns.items():
            matches = re.finditer(pattern, text, re.IGNORECASE)
            for match in matches:
                findings.append({
                    'type': pii_type,
                    'value': match.group(),
                    'start': match.start(),
                    'end': match.end()
                })
        
        return findings
    
    def mask_pii(self, text: str, mask_char: str = '*') -> Tuple[str, List[Dict]]:
        """
        Mask PII in text
        
        Args:
            text: Input text to mask
            mask_char: Character to use for masking
            
        Returns:
            Tuple of (masked text, list of masked items)
        """
        findings = self.detect_pii(text)
        masked_text = text
        masked_items = []
        
        # Sort findings by position (reverse order to maintain positions)
        findings.sort(key=lambda x: x['start'], reverse=True)
        
        for finding in findings:
            original_value = finding['value']
            
            # Different masking strategies for different types
            if finding['type'] == 'email':
                # Mask everything except domain
                parts = original_value.split('@')
                if len(parts) == 2:
                    masked_value = f"{mask_char * 5}@{parts[1]}"
                else:
                    masked_value = mask_char * len(original_value)
            elif finding['type'] in ['ssn', 'credit_card']:
                # Show only last 4 digits
                masked_value = mask_char * (len(original_value) - 4) + original_value[-4:]
            elif finding['type'] == 'phone':
                # Mask all but last 4 digits
                digits_only = re.sub(r'\D', '', original_value)
                if len(digits_only) >= 4:
                    masked_value = mask_char * (len(original_value) - 4) + original_value[-4:]
                else:
                    masked_value = mask_char * len(original_value)
            else:
                # Full mask for other types
                masked_value = mask_char * len(original_value)
            
            # Replace in text
            masked_text = masked_text[:finding['start']] + masked_value + masked_text[finding['end']:]
            
            masked_items.append({
                'type': finding['type'],
                'original': original_value,
                'masked': masked_value
            })
        
        return masked_text, masked_items
    
    def mask_client_info(self, text: str, client_name: str = None) -> str:
        """
        Mask client-specific information
        
        Args:
            text: Input text
            client_name: Client name to mask (if known)
            
        Returns:
            Text with client info masked
        """
        masked_text = text
        
        # If client name is provided, mask it
        if client_name:
            masked_text = re.sub(
                re.escape(client_name),
                '[CLIENT_NAME]',
                masked_text,
                flags=re.IGNORECASE
            )
        
        # Mask common client-related terms
        client_terms = ['Company', 'Corporation', 'Corp', 'Inc', 'LLC', 'Ltd']
        for term in client_terms:
            pattern = r'\b[A-Z][A-Za-z0-9]+\s+' + term + r'\b'
            masked_text = re.sub(pattern, '[ORGANIZATION]', masked_text)
        
        return masked_text

File: pii-cleaner/src/utils/piiDetector.js
// PII Detection and Masking Utilities

/**
 * Detects and masks various types of PII in text
 */
export class PIIDetector {
  constructor() {
    // Regular expressions for different PII types
    this.patterns = {
      email: /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g,
      phone: /(\+\d{1,3}[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b/g,
      ssn: /\b\d{3}-\d{2}-\d{4}\b/g,
      creditCard: /\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b/g,
      // Common name patterns (simplified)
      name: /\b[A-Z][a-z]+ [A-Z][a-z]+\b/g,
      ipAddress: /\b(?:\d{1,3}\.){3}\d{1,3}\b/g,
      url: /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_+.~#?&//=]*)/g,
      // AWS Access Key patterns
      awsAccessKey: /\b(A3T[A-Z0-9]|AKIA|AGPA|AIDA|AROA|AIPA|ANPA|ANVA|ASIA)[A-Z0-9]{16}\b/g,
      // AWS Secret Key pattern
      awsSecretKey: /\b[A-Za-z0-9/+=]{40}\b/g,
    };
  }

  /**
   * Detects all PII in the given text
   * @param {string} text - Text to analyze
   * @returns {Object} - Detected PII grouped by type
   */
  detectPII(text) {
    const detected = {};
    
    for (const [type, pattern] of Object.entries(this.patterns)) {
      const matches = text.match(pattern);
      if (matches && matches.length > 0) {
        detected[type] = [...new Set(matches)]; // Remove duplicates
      }
    }
    
    return detected;
  }

  /**
   * Masks PII in the given text
   * @param {string} text - Text to mask
   * @param {Array} typesToMask - Types of PII to mask (default: all)
   * @returns {string} - Masked text
   */
  maskPII(text, typesToMask = null) {
    let maskedText = text;
    const types = typesToMask || Object.keys(this.patterns);
    
    types.forEach(type => {
      if (this.patterns[type]) {
        maskedText = maskedText.replace(this.patterns[type], (match) => {
          return this.getMaskForType(type, match);
        });
      }
    });
    
    return maskedText;
  }

  /**
   * Gets appropriate mask for PII type
   * @param {string} type - Type of PII
   * @returns {string} - Masked text
   */
  getMaskForType(type) {
    const masks = {
      email: '[EMAIL REDACTED]',
      phone: '[PHONE REDACTED]',
      ssn: '[SSN REDACTED]',
      creditCard: '[CARD REDACTED]',
      name: '[NAME REDACTED]',
      ipAddress: '[IP REDACTED]',
      url: '[URL REDACTED]',
      awsAccessKey: '[AWS_KEY REDACTED]',
      awsSecretKey: '[AWS_SECRET REDACTED]',
    };
    
    return masks[type] || '[REDACTED]';
  }

  /**
   * Generates a report of detected PII
   * @param {Object} detected - Detected PII
   * @returns {string} - Formatted report
   */
  generateReport(detected) {
    let report = 'PII Detection Report:\n\n';
    
    if (Object.keys(detected).length === 0) {
      report += 'No PII detected.\n';
      return report;
    }
    
    for (const [type, items] of Object.entries(detected)) {
      report += `${type.toUpperCase()}:\n`;
      items.forEach(item => {
        report += `  - ${item}\n`;
      });
      report += '\n';
    }
    
    return report;
  }
}
